### PR TITLE
Jagged Arbitrary Masked Self Attention support 

### LINKED ISSUE #276: [FEA] Jagged Arbitrary Masked Self Attention support
mcore/te does not support jagged attention with  **Arbitrary Masked**  (It only supports causal). But in our scenario, this is a desirable feature. In our sid_gr example, we now use padded dense with arbitrary mask which is memory and compute intensive. 

**Describe the solution you'd like**

We can either wait for FA3 with arbitrary mask or write our only implementation. And then adapt it into sid example.

### PR DESCRIPTION
**Pull request by z52527**
Resolves #276

### Summary

Replace Megatron-Core's `TransformerBlock` (padded dense + `DotProductAttention`) with a `JaggedTransformerBlock` that flattens all batch sequences into a single sequence (B=1) and uses jiayus's Flash Attention with `arbitrary_func` mask encoding. Zero padding, block sparsity skips cross-batch blocks.

Both backends are preserved via `use_jagged_flash_attn` flag; mask / `arbitrary_func` is caller-provided, not built inside the model.

### Architecture

```
Caller builds arbitrary_func (or attention_mask)
         │
         ▼
SIDGRModel.forward / generate
         │
         ▼
decoder_step ── routes by use_jagged_flash_attn ──┐
         │                                         │
    FA path                                   mcore path
         │                                         │
JaggedTransformerBlock                    TransformerBlock
  unsqueeze(0) → [1,T,D]                pad → [S,B,D]
  flash_attn_func(arbitrary=True)        DotProductAttention
  squeeze(0) → [T,D]                    unpad → jagged
```

### Changed Files

| File | What Changed |
|------|-------------|
| `model/jagged_flash_attn_block.py` | **New.** `JaggedGPTLayer` (FA kernel per layer), `JaggedFlashAttnBlock` (layer stack), `JaggedTransformerBlock` (wrapper: unsqueeze→FA→squeeze) |
| `model/attention_mask.py` | **New functions.** `dense_mask_to_arbitrary_func` (per-batch dense→interval encoding), `build_jagged_causal_arbitrary_func` (vectorized, offsets→flattened causal arbitrary_func) |
| `model/gpt_model.py` | `SIDGRDecoder`: `use_jagged_flash_attn` flag selects `JaggedTransformerBlock` or mcore `TransformerBlock`. `decoder_step`: pure routing, zero mask construction. `forward/generate`: caller passes `arbitrary_func` or `attention_mask` |
| `model/__init__.py` | `get_sid_gr_model`: propagate `use_jagged_flash_attn` flag |
| `tests/test_jagged_flash_attn_block.py` | **New.** Smoke / correctness / backward tests for `JaggedGPTLayer`; FA-vs-reference comparison for `JaggedTransformerBlock`; batch isolation test |
| `tests/test_dense_mask_to_arbitrary_func.py` | **New.** Roundtrip tests for `dense_mask_to_arbitrary_func`; flattened jagged causal / target-grouped tests; `dense_mask_to_jagged_arbitrary_func` as test utility |

### Test Plan

- [ ] `pytest examples/sid_gr/tests/test_dense_mask_to_arbitrary_func.py` — arbitrary_func encoding roundtrip
- [ ] `pytest examples/sid_gr/tests/test_jagged_flash_attn_block.py` — FA layer correctness, JaggedTransformerBlock vs padded reference, batch isolation
- [ ] `mypy examples/sid_gr/model/gpt_model.py` — type check clean

### CODE REVIEW

File: examples/sid_gr/tests/test_jagged_flash_attn_block.py [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+from jagged_flash_attn_block import (
+    JaggedFlashAttnBlock,
+    JaggedGPTLayer,
+    JaggedTransformerBlock,
+)
+
+sys.path.pop(0)
+
+try:
+    from flash_attn.cute.interface import flash_attn_func  # noqa: F401
```
Comment (shijieliu) [inline examples/sid_gr/tests/test_jagged_flash_attn_block.py:48]: do we need to add arbitrary mask flash atten as 3rd party and build it in dockerfile?

File: examples/sid_gr/model/gpt_model.py [hunk 2]
```diff
@@ -513,66 +529,39 @@ def decoder_step(
...
-            attention_mask=attention_mask,
-            packed_seq_params=packed_seq_params,  # we now enforce arbitrary attention mask + dense padding
-        )
-
-        if padding_to_dense:
-            output_hidden_states = _transpose_dense_to_jagged(
-                decoder_output_hidden_states,
-                input_offsets,
-                input_max_seqlen,
+            assert (
```
Comment (shijieliu) [inline examples/sid_gr/model/gpt_model.py]: do we need to maintain mcore path? how about just remove it? @JacoCheung @z52527

File: examples/sid_gr/model/attention_mask.py [hunk 3]
```diff
@@ -207,6 +207,120 @@ def padded_target_aware_causal_mask(
     return ~valid_attn_mask
 
 
+def dense_mask_to_arbitrary_func(
```
Comment (shijieliu) [inline examples/sid_gr/model/attention_mask.py:210]: where is this func used? better to make it a fused triton function for better perf

**Comment by shijieliu** [comment]
update arbitary mask FA in 3rd party module

File: examples/sid_gr/model/attention_mask.py (hunk 3 above)
Comment (z52527) [reply examples/sid_gr/model/attention_mask.py:210]: This func only used for testing.

File: examples/sid_gr/tests/test_jagged_flash_attn_block.py (hunk 1 above)
Comment (z52527) [reply examples/sid_gr/tests/test_jagged_flash_attn_block.py:48]: Done.

**Comment by JacoCheung** [comment]
:x: **Pipeline #49662233** -- failed

| Job | Status | Log |
|-----|--------|-----|
| pre_check | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616901) |
| train_build | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616902) |
| inference_build | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616903) |
| tritonserver_build | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616904) |
| build_whl | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616905) |
| dynamicemb_test_fwd_bwd_8gpus | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616906) |
| dynamicemb_test_load_dump_8gpus | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616907) |
| unit_test_1gpu_a100 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616908) |
| unit_test_1gpu_h100 | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616909) |
| unit_test_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616911) |
| unit_test_tp_4gpu | :x: failed | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616912) |
| L20_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616913) |
| inference_unit_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616914) |
| inference_test_1gpu | :white_check_mark: success | [view](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/jobs/306616915) |

**Result: 9/14 jobs passed**

[View full pipeline](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/49662233)

**Comment by z52527** [comment]
track [CI](https://gitlab-master.nvidia.com/Devtech-Compute/distributed-recommender/-/pipelines/49663804) here

**Review by shijieliu — APPROVED** [review]
